src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -1384,35 +1551,74 @@ ExtractFirstDistributedTableId(Query *query)
 
 
 /*
- * ExtractPartitionValue extracts the partition column value from a the target
- * of an INSERT command. If a partition value is missing altogether or is
- * NULL, this function throws an error.
+ * ExtractInsertValuesList extracts the partition column value for an INSERT
+ * command and returns it within an InsertValues struct. For single-row INSERTs
+ * this is simply a value extracted from the target list, but multi-row INSERTs
+ * will generate a List of InsertValues, each with full row values in addition
+ * to the partition value. In this case, the range table index corresponding to
+ * the VALUES clause will also be returned in the varno output parameter. If a
+ * partition value is missing altogether or is NULL, this function errors.
  */
-static Expr *
-ExtractInsertPartitionValue(Query *query, Var *partitionColumn)
+static List *
+ExtractInsertValuesList(Query *query, Var *partitionColumn, Index *partitionvarno)
 {
+	List *insertValuesList = NIL;
 	TargetEntry *targetEntry = get_tle_by_resno(query->targetList,
 												partitionColumn->varattno);
+	*partitionvarno = 0; /* set to invalid RTE index by default */
+
 	if (targetEntry == NULL)
 	{
 		ereport(ERROR, (errcode(ERRCODE_NULL_VALUE_NOT_ALLOWED),
 						errmsg("cannot perform an INSERT without a partition column "
 							   "value")));
 	}
 
-	return targetEntry->expr;
+	/* We've got a multi-row INSERT. Extract all VALUES */
```

Comment:
Might help elaborating a bit on how multi-row INSERTs are represented in postgres.